File: todo-back/domain/repository/todo.go
package repository

import (
	"context"
	"todo-back/domain/model"
)

type TodoListRequest struct {
	Paging  *Paging
	Sorting *TodoSort
	Filter  *TodoFilter
}

type TodoListOutput struct {
	TodoList []model.Todo
	HasNext  bool
}

type TodoSort struct {
	CreateTime *SortField
	DoneTime   *SortField
}

type TodoFilter struct {
	ID          *FilterField[model.TodoID]
	Title       *FilterField[string]
	Description *FilterField[string]
	Status      *FilterField[model.TodoStatus]
}

type Todo interface {
	Create(ctx context.Context, todo model.Todo) (model.TodoID, error)
	Update(ctx context.Context, id model.TodoID, updatedTodo model.Todo) error
	Delete(ctx context.Context, id model.TodoID) error
	Get(ctx context.Context, id model.TodoID) (model.Todo, error)
	List(ctx context.Context, request TodoListRequest) (TodoListOutput, error)
}

File: todoapp/lib/domain/model/todo.dart
class Todo {
  final TodoId id;
  final String title;
  final String description;
  final DateTime createTime;
  final DateTime? doneTime;
  final TodoStatus status;

  Todo(
      {required this.id,
      required this.title,
      required this.description,
      required this.createTime,
      required this.doneTime,
      required this.status});
}

class TodoId {
  final String id;
  TodoId({required this.id});

  @override
  String toString() {
    return id;
  }
}

enum TodoStatus {
  done,
  scheduled,
  draft,
}
